quil-rs/src/waveform/templates.rs

```diff
@@ -0,0 +1,264 @@
+//! Templates for generating discrete IQ value sequences representing Quil's defined set of waveforms.
+
+use std::f64::consts::PI;
+
+use crate::{imag, real};
+use ndarray::Array;
+use num_complex::Complex64;
+use statrs::function::erf::erf;
+
+use serde::{Deserialize, Serialize};
+
+const J: Complex64 = Complex64::new(0f64, 1f64);
+
+/// Modulate and phase shift waveform IQ data in place.
+pub fn apply_phase_and_detuning(
+    iq_values: &mut [Complex64],
+    phase: f64,
+    detuning: f64,
+    sample_rate: f64,
+) {
+    for (index, value) in iq_values.iter_mut().enumerate() {
+        *value *= (2.0 * PI * J * (detuning * (index as f64) / sample_rate + phase)).exp();
+    }
+}
+
+/// A custom `ceil()` method that includes a machine-epsilon sized region above each integer in the
+/// set of values that are mapped to that integer. In other words:
+/// ceil_eps(x) = n, for all x and integers n s.t. n <= x < (n + n*epsilon)
+///
+/// To handle accumulated floating point errors in sweeps above typical floating point imprecision
+/// we make epsilon 10x larger than floating point epsilon.
+pub(super) fn ceiling_with_epsilon(value: f64) -> f64 {
+    let truncated = value - (value * 10.0 * std::f64::EPSILON);
+    truncated.ceil()
+}
+
+pub fn polar_to_rectangular(magnitude: f64, angle: crate::units::Radians<f64>) -> Complex64 {
+    magnitude * imag!(angle.0).exp()
+}
+
+#[derive(Clone, Debug)]
+pub struct BoxcarKernelParameters {
+    pub phase: crate::units::Cycles<f64>,
+    pub scale: f64,
+    pub sample_count: u64,
+}
+
+/// Return the IQ value used for a boxcar kernel
+pub fn boxcar_kernel(parameters: BoxcarKernelParameters) -> Complex64 {
+    polar_to_rectangular(
+        parameters.scale / parameters.sample_count as f64,
+        parameters.phase.into(),
+    )
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ErfSquareParameters {
+    /// Full duration of the pulse (s)
+    pub duration: f64,
+    /// Slope of erf shoulders (2x FWHM of erf in s)
+    pub risetime: f64,
+    /// Generate wavform samples at this rate (Hz)
+    pub sample_rate: f64,
+    /// Length of zero padding to add to beginning of pulse (s)
+    pub pad_left: f64,
+    /// Length of zero padding to add to end of pulse (s)
+    pub pad_right: f64,
+    /// Toggle for positive/negative polarity
+    pub positive_polarity: bool,
+    /// Scale to apply to waveform envelope
+    pub scale: f64,
+    /// Phase shift for entire waveform
+    pub phase: f64,
+    /// Explicit detuning to bake into iq values
+    pub detuning: f64,
+}
+
+/// Creates a waveform with a flat top and edges that are error functions (erfs).
+pub fn erf_square(parameters: ErfSquareParameters) -> Vec<Complex64> {
```

Comment:
I do wonder if these should be methods on the struct, rather than top level methods. Would be less import to worry about for users.

Reply:
yep, that was what I meant in the MR description. Moved them and added a unifying trait